tests/sentry/incidents/test_logic.py

```diff
@@ -852,6 +853,69 @@ def test_update_from_include_all(self):
             sub.project for sub in QuerySubscription.objects.filter(alert_rules=alert_rule)
         ] == [new_project]
 
+    @pytest.mark.chris
+    def test_with_attached_incident(self):
+        # A snapshot of the pre-updated rule should be created, and the incidents should also be resolved.
+        incident = self.create_incident()
+        incident.update(alert_rule=self.alert_rule)
+        incident_2 = self.create_incident()
+        incident_2.update(alert_rule=self.alert_rule)
+
+        # Give the rule some actions and triggers so we can verify they've been snapshotted correctly.
+        trigger = create_alert_rule_trigger(
+            self.alert_rule, "hello", AlertRuleThresholdType.ABOVE, 1000, 400
+        )
+        create_alert_rule_trigger_action(
+            trigger,
+            AlertRuleTriggerAction.Type.EMAIL,
+            AlertRuleTriggerAction.TargetType.USER,
+            target_identifier=six.text_type(self.user.id),
+        )
+        trigger_count = AlertRuleTrigger.objects.all().count()
+        action_count = AlertRuleTriggerAction.objects.all().count()
+
+        updated_projects = [self.project, self.create_project(fire_project_created=True)]
+        updated_rule = update_alert_rule(
+            self.alert_rule,
+            projects=updated_projects,
+            query="level:warning",
+            aggregation=QueryAggregations.UNIQUE_USERS,
+            time_window=50,
+            threshold_period=2,
+        )
+
+        incident.refresh_from_db()
+        incident_2.refresh_from_db()
+        rule_snapshot = AlertRule.objects_with_deleted.filter(name=self.alert_rule.name).exclude(
+            id=updated_rule.id
+        )
+        assert rule_snapshot.count() == 1
+        rule_snapshot = rule_snapshot.first()
+
+        # Rule snapshot should have properties of the rule before it was updated.
+        assert rule_snapshot.id != updated_rule.id
+        assert rule_snapshot.name == updated_rule.name
+        assert rule_snapshot.query == "level:error"
+        assert rule_snapshot.time_window == 10
+        assert rule_snapshot.aggregation == QueryAggregations.TOTAL.value
+        assert rule_snapshot.threshold_period == 1
+
+        # Incidents should now be pointing to the rule snapshot.
+        assert incident.alert_rule.id == rule_snapshot.id
+        assert incident_2.alert_rule.id == rule_snapshot.id
+        assert incident.alert_rule.name == updated_rule.name
+        assert incident_2.alert_rule.name == updated_rule.name
+
+        # Incidents should be resolved
+        assert incident.status == IncidentStatus.CLOSED.value
+        assert incident_2.status == IncidentStatus.CLOSED.value
+
+        # Action and trigger counts should double (from 1 to 2)
+        assert AlertRuleTrigger.objects.all().count() == trigger_count * 2
+        assert AlertRuleTriggerAction.objects.all().count() == action_count * 2
+
+        # TODO: Verify actions and triggers have the same properties?
```

Comment:
Probably worth verifying that they have the same properties, and that they're not associated with the old alert rule.